Secure socks proxy: Modify how proxy options can be set

stephaniehingtgen opened the pull request:
**What this PR does / why we need it**:

This PR refactors how the proxy options are set. Perviously, it was set mainly in the `HTTPClientOptions()` function, but then the socks user was set in the client. This made it both harder to understand when and why things are being set, and prevented non-http clients from setting it up in a straightforward manner.

This PR extracts all the functionality into the function `ProxyOptions()`, which simplifies the code base and is non-http specific. 

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/hosted-grafana/issues/4030

stephaniehingtgen committed:
Secure socks proxy: update where username is set for proxy

stephaniehingtgen committed:
Move into a new function

stephaniehingtgen committed:
Add unit tests, clean up

andresmgot commented on backend/common.go:
```diff
@@ -122,6 +123,11 @@ func (s *DataSourceInstanceSettings) HTTPClientOptions() (httpclient.Options, er
 
 	setCustomOptionsFromHTTPSettings(&opts, httpSettings)
 
+	opts.ProxyOptions, err = s.ProxyOptions()
```
I believe you need to do the same for `AppInstanceSettings` since now App plugins won't read the proxy configuration (which was done before in the function `parseHTTPSettings`)

stephaniehingtgen replied:
I was looking through an example where this is used for an app plugin, and I don't think this is needed, since this functionality can only live on the datasource part of the plugin. In the zabbix plugin, for example, it uses the [data souce instance settings](https://github.com/alexanderzobnin/grafana-zabbix/blob/df0ef4ba6ec3fa439a2138cb876cc23871639d90/pkg/datasource/datasource.go#L32), which is where PDC is then enabled from today.

andresmgot replied:
okay, I am just pointing the difference, if this is a feature only available for datasource plugins (or using datasource settings) then :shrug:

stephaniehingtgen replied:
For sure, I really appreciate you pointing it out! It is just a feature for datasource plugins/datasource settings :)

andresmgot reviewed (commented):
> Additionally, it removes the function SecureSocksProxyEnabledOnDS(). I'm not sure if this is allowed or how deprecation should work for this, so let me know if I should change it. It is currently used in 2 plugins, so I was planning on modifying them as soon as this is merged in.

It is allowed as long as you give people notice but I am not sure why you are removing it? Seems like it won't hurt to keep it, maybe adding a log line with a warning suggesting a better the recommended approach would be better?

stephaniehingtgen committed:
Update per PR comments

stephaniehingtgen commented on backend/common.go:
```diff
@@ -214,3 +220,55 @@ func propagateTenantIDIfPresent(ctx context.Context) context.Context {
+		opts.Enabled = true
+	} else {
+		return opts, nil
+	}
```
Updated :) I added `SecureSocksProxyEnabledOnDS` back in, but it returns nil if false

stephaniehingtgen commented:
> It is allowed as long as you give people notice but I am not sure why you are removing it?

I had removed it because it felt a bit weird to me having two sections of code looking at the json data, but it definitely doesn't need to be removed. I added it back in - let me know what you think :)

andresmgot reviewed (approved):
LGTM

andresmgot commented:
> > It is allowed as long as you give people notice but I am not sure why you are removing it?
> 
> I had removed it because it felt a bit weird to me having two sections of code looking at the json data, but it definitely doesn't need to be removed. I added it back in - let me know what you think :)

Feel free to deprecate it or suggest a better way to do it in the code
